flink-table/flink-table-api-bridge-common/src/main/java/org/apache/flink/table/api/bridge/internal/BaseStreamTableEnvironmentImpl.java

```diff
@@ -0,0 +1,329 @@
+                    String.format(
+                            "Expected a single transformation for query: %s\n Got: %s",
+                            table.getQueryOperation().asSummaryString(), transformations));
+        }
+
+        return (Transformation<T>) transformations.get(0);
+    }
+
+    protected <T> DataType wrapWithChangeFlag(TypeInformation<T> outputType) {
+        TupleTypeInfo tupleTypeInfo =
+                new TupleTypeInfo<Tuple2<Boolean, T>>(Types.BOOLEAN(), outputType);
+        return TypeConversions.fromLegacyInfoToDataType(tupleTypeInfo);
+    }
+
+    protected <T> DataStreamQueryOperation<T> asQueryOperation(
+            DataStream<T> dataStream, Optional<List<Expression>> fields) {
+        TypeInformation<T> streamType = dataStream.getType();
+
+        // get field names and types for all non-replaced fields
+        FieldInfoUtils.TypeInfoSchema typeInfoSchema =
+                fields.map(
+                                f -> {
+                                    FieldInfoUtils.TypeInfoSchema fieldsInfo =
+                                            FieldInfoUtils.getFieldsInfo(
+                                                    streamType, f.toArray(new Expression[0]));
+
+                                    // check if event-time is enabled
+                                    validateTimeCharacteristic(fieldsInfo.isRowtimeDefined());
+                                    return fieldsInfo;
+                                })
+                        .orElseGet(() -> FieldInfoUtils.getFieldsInfo(streamType));
+
+        return new DataStreamQueryOperation<>(
+                dataStream, typeInfoSchema.getIndices(), typeInfoSchema.toResolvedSchema());
+    }
+
+    protected void validateTimeCharacteristic(boolean isRowtimeDefined) {
+        if (isRowtimeDefined
+                && executionEnvironment.getStreamTimeCharacteristic()
+                        != TimeCharacteristic.EventTime) {
+            throw new ValidationException(
+                    String.format(
+                            "A rowtime attribute requires an EventTime time characteristic in stream environment. But is: %s",
+                            executionEnvironment.getStreamTimeCharacteristic()));
+        }
+    }
+
+    @Override
+    protected QueryOperation qualifyQueryOperation(
+            ObjectIdentifier identifier, QueryOperation queryOperation) {
+        if (queryOperation instanceof DataStreamQueryOperation) {
+            DataStreamQueryOperation<?> operation = (DataStreamQueryOperation) queryOperation;
+            return new DataStreamQueryOperation<>(
+                    identifier,
+                    operation.getDataStream(),
+                    operation.getFieldIndices(),
+                    operation.getResolvedSchema());
+        } else {
+            return queryOperation;
+        }
+    }
+
+    public void attachAsDataStream(List<ModifyOperation> modifyOperations) {
```

Comment:
the old one was default scoped

Reply:
Same reason as this one https://github.com/apache/flink/pull/18065#discussion_r766516674